### PR TITLE
Carryall edge spawn & Harvester delivery

### PR DESCRIPTION
**Pull request by MatthijsBenschop**
This spawns the Carryall under either fog, shroud or -when that's not available- at the map edge.

It has already been debated whether that's the way to go, but looking at original d2k I really believe this is the best way. 

Next PR would be harvester insurance (in case of worm frenzy)

### CODE REVIEW

**Comment by Phrohdoh** [comment]
Off topic but a nit: AutoCaryall should not be named so if you can configure it to not be automatic.

Inverting that logic to accompany the name Caryall would be better.

**Comment by MatthijsBenschop** [comment]
@Phrohdoh It's still an automated unit, it just doesn't get new orders automatically.

**Comment by pchote** [comment]
I remember making the same comment somewhere else.  We're dealing with trait level names, and so if the trait is relying on another trait to give it orders then it is no longer automatic.

File: OpenRA.Mods.D2k/Activities/DeliverUnit.cs [hunk 1]
```diff
@@ -28,24 +30,25 @@ public class CarryUnit : Activity
 		readonly IPositionable positionable;
 		readonly IFacing cFacing; // Carryable facing
 		readonly IFacing sFacing; // Self facing
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Activities/DeliverUnit.cs]: Instead of having these comments I am fine with `selfFacing` and `carryableFacing`.

File: OpenRA.Mods.D2k/Activities/PickupUnit.cs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+using OpenRA.Mods.RA.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Activities
+{
+	public class PickupUnit : Activity
+	{
+		readonly Actor carryable;
+		readonly IMove movement;
+		readonly Carryable c;
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Activities/PickupUnit.cs]: Having a variable `carryable` as a reference to an Actor type, while having a reference to a Carryable type isn't good design IMO.

I suggest `Actor toPickup` and `Carryable carryable`.

File: OpenRA.Mods.D2k/Traits/AutoCarryall.cs [hunk 3]
```diff
@@ -30,9 +34,12 @@ public class AutoCarryall : INotifyBecomingIdle, INotifyKilled, ISync, IRender
 	{
 		readonly Actor self;
 		readonly WRange carryHeight;
+		readonly AutoCarryallInfo info;
 
 		// The actor we are currently carrying.
-		[Sync] Actor carrying;
+		[Sync] public Actor Carrying { get; internal set; }
+
+		public bool HasCarryableAttached { get; internal set; }
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/AutoCarryall.cs]: This should be ` { get { return Carrying != null; } }`.

File: OpenRA.Mods.D2k/Traits/AutoCarryall.cs [hunk 4]
```diff
@@ -148,6 +165,10 @@ public void Killed(Actor self, AttackInfo e)
 		// Called when carryable is inside.
 		public void AttachCarryable(Actor carryable)
 		{
+			HasCarryableAttached = true;
+			Busy = true; // necesary when not in automatic mode
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/AutoCarryall.cs]: If this comment is to remain 'necessary' needs to be corrected.

File: OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+		public readonly string Actor = null;
+
+		[Desc("What the unit should start doing. Warning: If this is not a harvester", "it will break if you use FindResources.")]
+		public readonly string InitialActivity = null;
+
+		[Desc("Offset relative to structure-center in 2D (e.g. 1, 2)")]
+		public readonly CVec DeliveryOffset = CVec.Zero;
+
+		[ActorReference]
+		[Desc("Name AutoCarrier actor. This actor must have the AutoCarryall trait")]
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs]: `Name AutoCarrier actor` makes no sense to me.

File: OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+	public class DeliverFreeActor
+	{
+		readonly Actor self;
+
+		public DeliverFreeActor(Actor self, DeliverFreeActorInfo info)
+		{
+			this.self = self;
+
+			// Get a carryall spanw location
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs]: `spawn`

File: OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+				new LocationInit(location),
+				new CenterPositionInit(spawn),
+				new OwnerInit(self.Owner),
+				new FacingInit(initialFacing)
+			});
+
+			// Create harvester actor
+			var harv = self.World.CreateActor(false, info.Actor, new TypeDictionary
+			{
+				new LocationInit(location), // Redundant
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs]: If this is redundant (unnecessary?) don't include it.

File: OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+				new OwnerInit(self.Owner),
+			});
+
+			DoHarvesterDelivery(self.Location + info.DeliveryOffset, harv, carrier, info.InitialActivity);
+		}
+
+		public void DoHarvesterDelivery(CPos location, Actor client, Actor carrier, string clientInitialActivity)
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs]: `s/client/carried`

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+		public override bool Produce(Actor self, ActorInfo producee, string raceVariant)
+		{
+			// Get a spawn cell
+			var location = GetNearestInvisibleCell(self.Owner, self.Location);
+			var pos = self.World.Map.CenterOfCell(location);
+
+			var initialFacing = self.World.Map.FacingBetween(location, self.Location, 0);
+
+			// If aircraft, spawn at cruise altitude
+			var ai = producee.Traits.WithInterface<AircraftInfo>().First();
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: You need to use `FirstOrDefault` to properly conduct your null check.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+			// Get the closest shrouded cell
+			if (player.Shroud.ShroudEnabled && player.World.Type == WorldType.Regular)
+			{
+				var location = player.World.Map.Cells.Where(c => CheckUnexplored(c, checkplayers) == true).OrderBy(c => (c - toCell).LengthSquared).FirstOrDefault();
+
+				if (location != CPos.Zero)
+					return location;
+			}
+
+			// Last resort
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: This is understood as the last attempt to give a valid location (as it is the final return statement).

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+			return player.World.Map.ChooseRandomEdgeCell(player.World.SharedRandom);
+		}
+
+		// True when none of the players can see the cell
+		static bool CheckFogged(CPos c, IEnumerable<Player> players)
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: I would prefer `CellVisibleToAllPlayers` it is more verbose but also explains what to expect.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+		{
+			foreach (var p in players)
+				if (p.Shroud.IsVisible(c))
+					return false;
+
+			return true;
+		}
+
+		// True when none of the players have explored the cell
+		static bool CheckUnexplored(CPos c, IEnumerable<Player> players)
```
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: `CellExploredByAllPlayers` which brings an inversion of logic.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs (hunk 11 above)
Comment (penev92) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: I think we care whether it is INvisible to all players. Or visible to any.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs (hunk 12 above)
Comment (penev92) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: Same as the above comment.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs (hunk 11 above)
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: `if (!foo)`

Comment (penev92) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: There is still a difference between `Any` and `!All`.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+
+			return true;
+		}
+
+		// Gives a (1) cell that is invisible due to fog, or if none (2) a cell that has not been explored yet (shrouded) or as a last resort (3) a cell on map edge.
+		public static CPos GetNearestInvisibleCell(Player player, CPos toCell)
+		{
+			// Get list of players against wich a shroud/fog test should be done
+			var checkplayers = player.World.Players
+				.Where(p => p.Playable == true);
```
Comment (ScottNZ) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: The `== true` here is redundant.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+			}
+
+			// Last resort
+			return player.World.Map.ChooseRandomEdgeCell(player.World.SharedRandom);
+		}
+
+		// True when none of the players can see the cell
+		static bool CheckFogged(CPos c, IEnumerable<Player> players)
+		{
+			foreach (var p in players)
```
Comment (ScottNZ) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: players.Any(...) is preferred here.

File: mods/d2k/rules/aircraft.yaml [hunk 15]
```diff
@@ -29,6 +29,14 @@
 	-Selectable:
 	AutoCarryall:
 
+CARRYALL.reinforce:
+	Inherits: ^CARRYALL
+	-Buildable
```
Comment (ScottNZ) [inline mods/d2k/rules/aircraft.yaml]: Missing colon.

File: OpenRA.Mods.D2k/Activities/PickupUnit.cs [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+					}
+
+				case State.MoveToCarryable: // We arrived, move on top
+
+					if (self.Location == carryable.Location)
+					{
+						state = State.Turn;
+						return this;
+					}
+					else
```
Comment (ScottNZ) [inline OpenRA.Mods.D2k/Activities/PickupUnit.cs]: Don't need this else; the `return this;` statement above removes the need for it.

File: OpenRA.Mods.D2k/Activities/PickupUnit.cs [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+						return Util.SequenceActivities(movement.MoveTo(carryable.Location, 0), this);
+
+				case State.Turn: // Align facing and Land
+
+					if (sFacing.Facing != cFacing.Facing)
+						return Util.SequenceActivities(new Turn(self, cFacing.Facing), this);
+					else
```
Comment (ScottNZ) [inline OpenRA.Mods.D2k/Activities/PickupUnit.cs]: Don't need this else; the `return this;` statement above removes the need for it.

File: OpenRA.Mods.D2k/Activities/PickupUnit.cs [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+					return Util.SequenceActivities(movement.MoveWithinRange(Target.FromActor(carryable), WRange.FromCells(4)), this);
+
+				case State.LockCarryable:
+					// Last check
+					if (c.StandbyForPickup(self))
+					{
+						state = State.MoveToCarryable;
+						return this;
+					} 
+					else
```
Comment (ScottNZ) [inline OpenRA.Mods.D2k/Activities/PickupUnit.cs]: Don't need this else; the `return this;` statement above removes the need for it.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+				if (p.Shroud.IsVisible(c))
+					return false;
+
+			return true;
+		}
+
+		// True when none of the players have explored the cell
+		static bool CheckUnexplored(CPos c, IEnumerable<Player> players)
+		{
+			foreach (var p in players)
```
Comment (ScottNZ) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: players.Any(...) (or All) is preferred here.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs (hunk 11 above)
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: The current implementation is `false if !any` so changing it to `true if all` isn't hard.

**Comment by Mailaender** [comment]
Also needs a rebase now.

File: OpenRA.Mods.D2k/Traits/AutoCarryall.cs (hunk 3 above)
Comment (MatthijsBenschop) [inline OpenRA.Mods.D2k/Traits/AutoCarryall.cs]: No, because Carrying gets set when the carryable gets reserved - its not necessarily attached.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs (hunk 9 above)
Comment (MatthijsBenschop) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: Good catch :)

File: OpenRA.Mods.D2k/Activities/DeliverUnit.cs (hunk 1 above)
Comment (MatthijsBenschop) [inline OpenRA.Mods.D2k/Activities/DeliverUnit.cs]: Not part of this PR but will do

File: OpenRA.Mods.D2k/Traits/AutoCarryall.cs (hunk 3 above)
Comment (Phrohdoh) [inline OpenRA.Mods.D2k/Traits/AutoCarryall.cs]: Carrying means it is in the act of being carried, unless we are carrying without being attached.

Comment (MatthijsBenschop) [inline OpenRA.Mods.D2k/Traits/AutoCarryall.cs]: Okay so the variable name is wrong, but the logic isn't

Comment (penev92) [inline OpenRA.Mods.D2k/Traits/AutoCarryall.cs]: Actually I prefer `HasCarryableAttacked` over `Carrying`. The latter doesn't say much and doesn't sound as good. I don't think a property name that explains *too well* what it's for should be a problem.

**Comment by Mailaender** [comment]
This has been silently updated.

File: OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.D2k.Activities;
+using OpenRA.Mods.RA.Activities;
+using OpenRA.Mods.RA.Traits;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	[Desc("Delivers a spice harvester after construction of this building. Optionally provides the harvester insurance function.")]
```
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs]: `Optionally provides the harvester insurance function.` is not quite there.

File: OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs (hunk 11 above)
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/ProductionFromMapEdge.cs]: `Any` is a lot more performant, because it can abort earlier.

**Comment by Mailaender** [comment]
> This spawns the Carryall under either fog, shroud or -when that's not available- at the map edge.
>
> It has already been debated whether that's the way to go, but looking at original d2k I really believe this is the best way.

This will look weird and unpolished in replays with carryalls popping out of nowhere. Also the shellmap shroud hack indicates that this behavior although it tries to imitate the original 1:1 is not ideal for this engine.

File: OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs (hunk 20 above)
Comment (Mailaender) [inline OpenRA.Mods.D2k/Traits/Buildings/DeliverFreeActor.cs]: `Delivers a spice harvester after construction of this building.` should probably be `Delivers a unit` as this is far more general.

**Comment by Mailaender** [comment]
Otherwise this work really well! :+1:

**Comment by penev92** [comment]
I agree that map-edge-spawining would look better for both players and spectators, and also would be *a lot* easier on the CPU

**Comment by obrakmann** [comment]
While I agree that only spawning at map edges would look better, it would also be bad for balance in some cases, especially in the early game when players would possibly have to wait considerably longer for their first harvester than their opponents.  On maps such as "Imperial Basin" this would seriously handicap players that spawn in the middle. On the other hand, Imperial Basin is the only map we ship where that would really be an issue. Anyway, at the very least I think this aspect should be considered before asking for changes here.  Let's find a consensus.

**Comment by MatthijsBenschop** [comment]
Well.. I've already thought about making the under-shroud-spawn optional. Where normal maps will have edge spawn, but a map with custom yaml, or single player maps, could enable the option for shroud spawn. 

This would apply to carryall production, harvester delivery and harvester insurance.

b.t.w. I'll try to finish this pr asap.. just very limited in spare time :/

**Comment by penev92** [comment]
> Let's find a consensus.

My vote is for edge-spawning (actually I'm OK with spawning new carryalls from the High-tech factory) for harvester delivery for the reasons mentioned above, despite the slight advantage (a matter of seconds) that would give some players on **a couple** of maps.

**Comment by Unit158** [comment]
It may be best to give carryalls a speedy spawn (much faster than missile speed, like spy plane) so the advantage is nullified. Either that or make them impossible to target, as was the case in Dune 2. The biggest potential problems with edge spawn are tanks or turrets sniping spawning carryalls.

**Comment by MicroJOo** [comment]
im fine with edge-spawning and the option for shroud-spawning 

but i dont like carryall to spawn from a hightech factory ... or a speed-buff for spawning

**Comment by Mailaender** [comment]
Needs a rebase.

**Comment by penev92** [comment]
I'd really like to see this in the next playtest, as we really need the harvester insurance.

**Comment by pchote** [comment]
Activity has stalled here, so moving to the future milestone and closing for now.

**Comment by penev92** [comment]
This has been superseded, so I'm removing the `Future` milestone.
